build: Build minisketch test in `make check`, not in `make`

hebasto opened the pull request:
On master (d1e42659bbdd8da170542d8c638242cd94f71a7d):
```
$ ./autogen.sh && ./configure --without-gui --disable-wallet && make clean
$ make 2>&1 | grep LD | grep -v .la
  CXXLD    bitcoind
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoin-util
  CXXLD    test/test_bitcoin
  CXXLD    bench/bench_bitcoin
  CXXLD    minisketch/test
  CXXLD    test/fuzz/fuzz
  CXXLD    univalue/test/object
  CXXLD    univalue/test/unitester
$ make check 2>&1 | grep LD
  CCLD     exhaustive_tests
  CCLD     tests
```

With this PR:
```
$ ./autogen.sh && ./configure --without-gui --disable-wallet && make clean
$ make 2>&1 | grep LD | grep -v .la
  CXXLD    bitcoind
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoin-util
  CXXLD    test/test_bitcoin
  CXXLD    bench/bench_bitcoin
  CXXLD    test/fuzz/fuzz
  CXXLD    univalue/test/object
  CXXLD    univalue/test/unitester
$ make check 2>&1 | grep LD
  CXXLD    minisketch/test
  CCLD     exhaustive_tests
  CCLD     tests
```

In fact, this PR restores behavior that was before bitcoin/bitcoin#22646, and that behavior looks more optimal.

As an outcome, the `contrib/guix/libexec/build.sh` does not spend resources to build binaries which are not a part of the release package.

jonatack reviewed (commented):
Concept ACK modulo Win64 issue. Same result as pull description using debian 5.15.5-1 (2021-11-26), clang 14.

master
```
(master)$ make 2>&1 | grep LD
  CXXLD    libunivalue.la
  CCLD     libsecp256k1.la
  CXXLD    bitcoind
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoin-util
  CXXLD    test/test_bitcoin
  CXXLD    bench/bench_bitcoin
  CXXLD    minisketch/test
  CXXLD    test/fuzz/fuzz
  CXXLD    univalue/test/object
  CXXLD    univalue/test/unitester
  CXXLD    univalue/test/no_nul
  CXXLD    libbitcoinconsensus.la
(master)$ make check 2>&1 | grep LD
  CCLD     tests
  CCLD     valgrind_ctime_test
  CCLD     exhaustive_tests
```
this branch
```
((HEAD detached from origin/pr/23670))$ make 2>&1 | grep LD
  CXXLD    libunivalue.la
  CCLD     libsecp256k1.la
  CXXLD    bitcoind
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoin-util
  CXXLD    test/test_bitcoin
  CXXLD    bench/bench_bitcoin
  CXXLD    test/fuzz/fuzz
  CXXLD    libbitcoinconsensus.la
((HEAD detached from origin/pr/23670))$ make check 2>&1 | grep LD
  CXXLD    minisketch/test
  CXXLD    univalue/test/object
  CXXLD    univalue/test/unitester
  CXXLD    univalue/test/no_nul
  CCLD     tests
  CCLD     valgrind_ctime_test
  CCLD     exhaustive_tests
```

laanwj commented:
I think there's a slight value in always building all binaries (that are enabled by configure).

But as `src/test/test_bitcoin` `bench/bench_bitcoin` and `test/fuzz/fuzz`  are still always built, and you're only referring to test binaries of included dependencies, I'm ok with this change.

jamesob commented:
Concept ACK given, as @laanwj said, this is only omitting dependency-test compilation during `make`.

hebasto commented:
Closing. Seeing no straightforward way to integrate this PR into our CI setup.

hebasto committed:
build: Build minisketch test in `make check`, not in `make`

hebasto commented:
> Concept ACK modulo Win64 issue.

Fixed.

hebasto commented:
Rebased c39c553eb6d308a420e49f68e469cb88c673e5d0 -> 1593837ecbadea1fa8395e9de14e8f3ad3c0bf91 ([pr23670.03](https://github.com/hebasto/bitcoin/commits/pr23670.03) -> [pr23670.04](https://github.com/hebasto/bitcoin/commits/pr23670.04)) due to the conflict with #24212.

hebasto commented:
Rebased 1593837ecbadea1fa8395e9de14e8f3ad3c0bf91 -> e9fa9acfff6d28ea974e205111a23305c91c14bf ([pr23670.04](https://github.com/hebasto/bitcoin/commits/pr23670.04) -> [pr23670.05](https://github.com/hebasto/bitcoin/commits/pr23670.05)) due to the conflict with #24322.

hebasto commented:
Rebased e9fa9acfff6d28ea974e205111a23305c91c14bf -> 04dae33cff761c9135454b4abb51749f3e2b5d83 ([pr23670.05](https://github.com/hebasto/bitcoin/commits/pr23670.05) -> [pr23670.06](https://github.com/hebasto/bitcoin/commits/pr23670.06)).

fanquake reviewed (approved):
ACK 04dae33cff761c9135454b4abb51749f3e2b5d83 - you could update the PR description to remove the no-longer-existant binaries.

hebasto commented on src/Makefile.test.include:
```diff
@@ -374,7 +374,7 @@ endif
 if ENABLE_TESTS
 UNIVALUE_TESTS = univalue/test/object univalue/test/unitester
-noinst_PROGRAMS += $(UNIVALUE_TESTS)
+check_PROGRAMS += $(UNIVALUE_TESTS)
```
@MarcoFalke Yes. it is. Why did you point at it?

hebasto replied:
Why?

`test_bitcoin` is a part of the distributed package, but univalue tests are not. Why build them in `contrib/guix/libexec/build.sh`?

MarcoFalke replied:
I wasn't aware this changes the guix build and the release tarball. Maybe mention it in the description?

hebasto replied:
This PR does not change the release tarball.

MarcoFalke replied:
I find it convenient to be notified of univalue build failures on `make` when developing univalue.

hebasto replied:
"univalue build failures" or "univalue tests build failures"?

MarcoFalke replied:
Both.

Did you measure how many milliseconds this shaves off of the guix build? I'd bet it is far less than 0.1% of overall build time.

Unrelated: It may overall be more useful for devs to introduce a `make check_quick` to only run the bitcoin core unit tests (without any secp exhaustive_tests)

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23670#issuecomment-1191281796).

kristapsk commented:
Concept ACK

hebasto commented:
Updated 04dae33cff761c9135454b4abb51749f3e2b5d83 -> 6d58117a31a88eec3f0a103f9d1fc26cf0b48348 ([pr23670.06](https://github.com/hebasto/bitcoin/commits/pr23670.06) -> [pr23670.07](https://github.com/hebasto/bitcoin/commits/pr23670.07)):

- addressed @MarcoFalke's [comments](https://github.com/bitcoin/bitcoin/pull/23670#discussion_r925641435)

TheCharlatan commented:
ACK 6d58117a31a88eec3f0a103f9d1fc26cf0b48348
